File: app/src/main/java/com/uccd3223/p1_chai_boon_hong_2206806/PlaceValueActivity.java
package com.uccd3223.p1_chai_boon_hong_2206806;

import android.os.Bundle;
import android.view.View;
import android.widget.GridLayout;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.android.flexbox.FlexboxLayout;
import com.google.android.material.card.MaterialCardView;
import com.uccd3223.p1_chai_boon_hong_2206806.util.ExerciseGeneratorUtil;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;

public class PlaceValueActivity extends BaseGameActivity {
    private static final String STATE_TENS = "place_value_tens";
    private static final String STATE_ONES = "place_value_ones";
    private static final String STATE_LAST_TOTAL = "place_value_last_total";
    private static final String STATE_OPTIONS = "place_value_options";
    private static final String STATE_QUESTION_LOCKED = "place_value_question_locked";

    private FlexboxLayout tensContainerLayout;
    private FlexboxLayout onesContainerLayout;
    private GridLayout optionsContainerLayout;
    private TextView tvTensCount;
    private TextView tvOnesCount;
    private TextView tvFeedback;
    private ExerciseGeneratorUtil.PlaceValueData currentData;
    private List<Integer> currentOptions = Collections.emptyList();
    private int lastTotal = -1;
    private boolean questionLocked;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_place_value);
        setupGameModeUI();
        initializeUI();
        if (savedInstanceState != null && savedInstanceState.containsKey(STATE_TENS)) {
            restoreQuestion(savedInstanceState);
        } else {
            loadNextQuestion();
        }
    }

    private void initializeUI() {
        tensContainerLayout = findViewById(R.id.tensContainerLayout);
        onesContainerLayout = findViewById(R.id.onesContainerLayout);
        optionsContainerLayout = findViewById(R.id.optionsContainerLayout);
        tvTensCount = findViewById(R.id.tvTensCount);
        tvOnesCount = findViewById(R.id.tvOnesCount);
        tvFeedback = findViewById(R.id.tvFeedback);
        findViewById(R.id.btnBack).setOnClickListener(view -> finish());
    }

    private void loadNextQuestion() {
        questionLocked = false;
        resetFeedback(tvFeedback);
        do {
            currentData = ExerciseGeneratorUtil.generatePlaceValue(9);
        } while (currentData.getTotal() == lastTotal);
        lastTotal = currentData.getTotal();
        currentOptions = ExerciseGeneratorUtil
                .generateRecognitionOptions(currentData.getTotal(), 99, 4)
                .getOptions();
        renderQuestion();
    }

    private void restoreQuestion(Bundle state) {
        currentData = new ExerciseGeneratorUtil.PlaceValueData(
                state.getInt(STATE_TENS),
                state.getInt(STATE_ONES)
        );
        lastTotal = state.getInt(STATE_LAST_TOTAL, currentData.getTotal());
        questionLocked = state.getBoolean(STATE_QUESTION_LOCKED);
        ArrayList<Integer> restoredOptions = state.getIntegerArrayList(STATE_OPTIONS);
        currentOptions = restoredOptions == null ? Collections.emptyList() : restoredOptions;
        if (currentOptions.size() != 4) {
            currentOptions = ExerciseGeneratorUtil
                    .generateRecognitionOptions(currentData.getTotal(), 99, 4)
                    .getOptions();
        }
        renderQuestion();
        if (questionLocked) {
            restoreCompletedAnswerChoice(
                    optionsContainerLayout,
                    currentData.getTotal(),
                    tvFeedback,
                    successMessage(),
                    this::loadNextQuestion
            );
        }
    }

    private void renderQuestion() {
        renderVisualUI();
        renderOptionsUI();
    }

    private void renderVisualUI() {
        tensContainerLayout.removeAllViews();
        onesContainerLayout.removeAllViews();
        tvTensCount.setText(getResources().getQuantityString(
                R.plurals.tens_count,
                currentData.getTens(),
                currentData.getTens()
        ));
        tvOnesCount.setText(getResources().getQuantityString(
                R.plurals.ones_count,
                currentData.getOnes(),
                currentData.getOnes()
        ));

        float density = getResources().getDisplayMetrics().density;
        int margin = (int) (3 * density);
        int tensWidth = (int) (18 * density);
        int tensHeight = (int) (118 * density);
        int oneSize = (int) (28 * density);

        for (int index = 0; index < currentData.getTens(); index++) {
            ImageView view = new ImageView(this);
            view.setImageResource(R.drawable.ic_tens_block);
            view.setScaleType(ImageView.ScaleType.FIT_XY);
            view.setContentDescription(getString(R.string.ten_block_description));
            view.setImportantForAccessibility(View.IMPORTANT_FOR_ACCESSIBILITY_YES);
            FlexboxLayout.LayoutParams params = new FlexboxLayout.LayoutParams(tensWidth, tensHeight);
            params.setMargins(margin, margin, margin, margin);
            view.setLayoutParams(params);
            tensContainerLayout.addView(view);
        }

        for (int index = 0; index < currentData.getOnes(); index++) {
            ImageView view = new ImageView(this);
            view.setImageResource(R.drawable.ic_ones_block);
            view.setScaleType(ImageView.ScaleType.FIT_XY);
            view.setContentDescription(getString(R.string.one_block_description));
            view.setImportantForAccessibility(View.IMPORTANT_FOR_ACCESSIBILITY_YES);
            FlexboxLayout.LayoutParams params = new FlexboxLayout.LayoutParams(oneSize, oneSize);
            params.setMargins(margin, margin, margin, margin);
            view.setLayoutParams(params);
            onesContainerLayout.addView(view);
        }
    }

    private void renderOptionsUI() {
        List<AnswerChoice> choices = new ArrayList<>();
        for (int option : currentOptions) {
            choices.add(new AnswerChoice(
                    option,
                    getString(R.string.number_value, option),
                    getString(R.string.answer_number, option)
            ));
        }
        renderAnswerChoices(optionsContainerLayout, choices, (choice, card) ->
                checkAnswer(choice.getValue(), card));
    }

    private void checkAnswer(int selectedOption, MaterialCardView card) {
        if (questionLocked || isGameOver) return;
        if (selectedOption == currentData.getTotal()) {
            questionLocked = true;
            completeAnswerChoice(
                    optionsContainerLayout,
                    card,
                    tvFeedback,
                    successMessage(),
                    this::loadNextQuestion
            );
        } else {
            retryAnswerChoice(card, tvFeedback, getString(R.string.place_value_retry));
        }
    }

    private String successMessage() {
        String placeParts = getString(
                R.string.number_words_pair,
                getResources().getQuantityString(
                        R.plurals.tens_count,
                        currentData.getTens(),
                        currentData.getTens()
                ),
                getResources().getQuantityString(
                        R.plurals.ones_count,
                        currentData.getOnes(),
                        currentData.getOnes()
                )
        );
        String numberWords = ExerciseGeneratorUtil.numberToWords(currentData.getTotal())
                .toLowerCase(Locale.getDefault());
        return getString(R.string.place_value_success, placeParts, numberWords);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        if (currentData != null) {
            outState.putInt(STATE_TENS, currentData.getTens());
            outState.putInt(STATE_ONES, currentData.getOnes());
        }
        outState.putInt(STATE_LAST_TOTAL, lastTotal);
        outState.putIntegerArrayList(STATE_OPTIONS, new ArrayList<>(currentOptions));
        outState.putBoolean(STATE_QUESTION_LOCKED, questionLocked);
        super.onSaveInstanceState(outState);
    }
}

File: app/src/main/java/com/uccd3223/p1_chai_boon_hong_2206806/PlaceValueActivity.kt
package com.uccd3223.p1_chai_boon_hong_2206806

import androidx.core.graphics.toColorInt
import android.os.Bundle
import android.widget.Button
import android.widget.GridLayout
import android.widget.ImageView
import android.widget.TextView
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import com.google.android.flexbox.FlexboxLayout
import com.google.android.material.card.MaterialCardView
import com.uccd3223.p1_chai_boon_hong_2206806.util.ExerciseGeneratorUtil

class PlaceValueActivity : BaseGameActivity() {

    private lateinit var visualContainerLayout: FlexboxLayout
    private lateinit var optionsContainerLayout: GridLayout
    private lateinit var tvFeedback: android.widget.TextView
    private lateinit var currentData: ExerciseGeneratorUtil.PlaceValueData
    private var lastTotal: Int = -1

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_place_value)
        
        setupGameModeUI()

        initializeUI()
        loadNextQuestion()
    }

    private fun initializeUI() {
        visualContainerLayout = findViewById(R.id.visualContainerLayout)
        optionsContainerLayout = findViewById(R.id.optionsContainerLayout)
        tvFeedback = findViewById(R.id.tvFeedback)
        
        findViewById<Button>(R.id.btnBack).setOnClickListener {
            finish()
        }
    }

    private fun loadNextQuestion() {
        visualContainerLayout.removeAllViews()
        optionsContainerLayout.removeAllViews()

        do {
            currentData = ExerciseGeneratorUtil.generatePlaceValue(9)
        } while (currentData.total == lastTotal)
        lastTotal = currentData.total

        renderVisualUI()
        renderOptionsUI()
    }

    private fun renderVisualUI() {
        val dpScale = resources.displayMetrics.density
        
        // Dynamically calculate unit size to fit within the 1:1 square border
        // A Tens block is 1 unit wide, 10 units tall. A Ones block is 1 unit wide, 1 unit tall.
        val totalBlocks = currentData.tens + currentData.ones
        val unitSizeDp = when {
            totalBlocks <= 6 -> 24
            totalBlocks <= 12 -> 20
            else -> 18
        }
        val marginDp = 4
        
        val margin = (marginDp * dpScale).toInt()
        val unitSize = (unitSizeDp * dpScale).toInt()
        val tensHeight = unitSize * 10

        // Tens
        for (i in 0 until currentData.tens) {
            val tensView = ImageView(this).apply {
                setImageResource(R.drawable.ic_tens_block)
                scaleType = ImageView.ScaleType.FIT_XY
                val newLayoutParams = FlexboxLayout.LayoutParams(unitSize, tensHeight).apply {
                    setMargins(margin, margin, margin, margin)
                }
                this.layoutParams = newLayoutParams
            }
            visualContainerLayout.addView(tensView)
        }

        // Ones
        for (i in 0 until currentData.ones) {
            val onesView = ImageView(this).apply {
                setImageResource(R.drawable.ic_ones_block)
                scaleType = ImageView.ScaleType.FIT_XY
                val newLayoutParams = FlexboxLayout.LayoutParams(unitSize, unitSize).apply {
                    setMargins(margin, margin, margin, margin)
                }
                this.layoutParams = newLayoutParams
            }
            visualContainerLayout.addView(onesView)
        }
    }

    private fun renderOptionsUI() {
        val optionsData = ExerciseGeneratorUtil.generateRecognitionOptions(currentData.total, 99, 4)

        val colors = listOf("#9C27B0", "#1368CE", "#D89E00", "#00BCD4")

        for ((index, option) in optionsData.options.withIndex()) {
            val optionBtn = layoutInflater.inflate(R.layout.item_answer_choicer, optionsContainerLayout, false) as MaterialCardView
            optionBtn.setCardBackgroundColor(colors[index].toColorInt())
            
            val tv = optionBtn.findViewById<TextView>(R.id.answerChoicerText)
            tv.text = getString(R.string.number_value, option)
            optionBtn.contentDescription = getString(R.string.answer_number, option)
            
            val layoutParams = GridLayout.LayoutParams(
                GridLayout.spec(GridLayout.UNDEFINED, 1f),
                GridLayout.spec(GridLayout.UNDEFINED, 1f)
            ).apply {
                width = 0
                height = ViewGroup.LayoutParams.WRAP_CONTENT
            }
            optionBtn.layoutParams = layoutParams
            
            optionBtn.setOnClickListener { checkAnswer(option, optionBtn) }
            optionsContainerLayout.addView(optionBtn)
        }
    }

    private fun checkAnswer(selectedOption: Int, card: MaterialCardView) {
        if (selectedOption == currentData.total) {
            card.setCardBackgroundColor("#66BB6A".toColorInt())
            for (i in 0 until optionsContainerLayout.childCount) {
                optionsContainerLayout.getChildAt(i).isEnabled = false
            }
            
            tvFeedback.setText(R.string.great_job)
            tvFeedback.setTextColor("#66BB6A".toColorInt())
            tvFeedback.visibility = android.view.View.VISIBLE
            
            tvFeedback.postDelayed({
                if (!isGameOver) {
                    onQuestionCompleted()
                    loadNextQuestion()
                    tvFeedback.visibility = android.view.View.INVISIBLE
                }
            }, 550)
        } else {
            card.setCardBackgroundColor("#EF5350".toColorInt())
            
            tvFeedback.setText(R.string.try_again)
            tvFeedback.setTextColor("#EF5350".toColorInt())
            tvFeedback.visibility = android.view.View.VISIBLE
            
            val shake = AnimationUtils.loadAnimation(this, R.anim.shake)
            card.startAnimation(shake)
        }
    }
}
